### PR TITLE
Add PF driver and LM commands

### PR DESCRIPTION
**Pull request by GavinLi-NV**
1. Add PF driver
2. Add Admin Q and message sending command API
3. Add LM commands for suspend/resume, save/restore state and save pending bytes

### CODE REVIEW

File: drivers/common/virtio/virtio_admin.h [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL IBM OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE. */
+#include <linux/types.h>
```
Comment (kailiangz1) [inline drivers/common/virtio/virtio_admin.h]: we should not rely on linux types, use DPDK types

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 2]
```diff
@@ -62,6 +64,7 @@ struct virtio_vdpa_priv {
 
 #define VIRTIO_VDPA_INTR_RETRIES_USEC 1000
 #define VIRTIO_VDPA_INTR_RETRIES 256
+#include <virtio_lm.h>
```
Comment (kailiangz1) [inline drivers/vdpa/virtio/virtio_vdpa.c]: this file had  included in line 14

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 3]
```diff
@@ -72,6 +75,16 @@ static TAILQ_HEAD(virtio_vdpa_privs, virtio_vdpa_priv) virtio_priv_list =
 						  TAILQ_HEAD_INITIALIZER(virtio_priv_list);
 static pthread_mutex_t priv_list_lock = PTHREAD_MUTEX_INITIALIZER;
 
+static struct virtio_vdpa_mi_ops mi_ops = {
+	.get_mi_by_bdf = NULL,
+};
+
+void
```
Comment (kailiangz1) [inline drivers/vdpa/virtio/virtio_vdpa.c]: static

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 4]
```diff
@@ -898,6 +933,10 @@ virtio_vdpa_dev_probe(struct rte_pci_driver *pci_drv __rte_unused,
 		return -rte_errno;
 	}
 
+	/* check pf_priv before use it, might be null if not set */
+	priv->pf_priv = pf_priv;
+	if (!priv->pf_priv)
+		DRV_LOG(WARNING, "PF was not set");
```
Comment (kailiangz1) [inline drivers/vdpa/virtio/virtio_vdpa.c]: return error if no pf
		DRV_LOG(ERR, "dev:%s no pf driver found",
						 devname);
		rte_errno = EINVAL;
		return -rte_errno;

File: drivers/vdpa/virtio_mi/lm.c [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+{
+	if (strcmp(value, "2") == 0)
+		*(int *)ret_val = 1;
+	else
+		*(int *)ret_val = 0;
+
+	return 0;
+}
+
+#define VIRTIO_ARG_VDPA 	  "vdpa"
```
Comment (kailiangz1) [inline drivers/vdpa/virtio_mi/lm.c]: we can move this marco to virtio_api.h, both pf driver and vpda driver use this parameter

File: drivers/vdpa/virtio_mi/lm.c [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	return ret;
+}
+
+static int
+virtio_vdpa_mi_dev_probe(struct rte_pci_driver *pci_drv __rte_unused,
+		struct rte_pci_device *pci_dev)
+{
+	char devname[RTE_DEV_NAME_MAX_LEN] = {0};
+	struct virtio_vdpa_pf_priv *priv = NULL;
+	int vdpa = 0, ret = 0;
```
Comment (kailiangz1) [inline drivers/vdpa/virtio_mi/lm.c]: just ret;
no need =0

File: drivers/vdpa/virtio_mi/lm.c [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+	priv->vfio_dev_fd = rte_intr_dev_fd_get(pci_dev->intr_handle);
+	if (priv->vfio_dev_fd < 0) {
+		DRV_LOG(ERR, "%s failed to get vfio dev fd", devname);
+		rte_errno = rte_errno ? rte_errno : EINVAL;
+		goto error;
+	}
+
+	virtio_pci_dev_features_get(priv->vpdev, &priv->guest_features);
+	if (!(priv->guest_features & (VIRTIO_VDPA_MI_SUPPORTED_FEATURE))) {
```
Comment (kailiangz1) [inline drivers/vdpa/virtio_mi/lm.c]: guest_features should name device_features

File: drivers/vdpa/virtio_mi/lm.c [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+			break;
+		}
+	}
+	pthread_mutex_unlock(&mi_priv_list_lock);
+
+	if (found) {
+		virtio_pci_dev_reset(priv->vpdev);
+
+		/* Tell the host we've noticed this device. */
+		virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_ACK);
```
Comment (kailiangz1) [inline drivers/vdpa/virtio_mi/lm.c]: remove needn't set ack and driver status

File: drivers/vdpa/virtio_mi/lm.c [hunk 9]
```diff
@@ -66,6 +65,214 @@ virtio_vdpa_get_mi_by_bdf(const char *bdf)
...
+}
+
+static int
+virtio_vdpa_admin_queue_alloc(struct virtio_vdpa_pf_priv *priv)
+{
+	struct virtio_hw *hw = &priv->vpdev->hw;
+	uint16_t i, queue_idx;
+	int ret;
+
+	hw->max_queue_pairs = priv->vpdev->common_cfg->num_queues / 2;
```
Comment (kailiangz1) [inline drivers/vdpa/virtio_mi/lm.c]: max_queue_pairs can be get from virtio_net_config, should not caculate like this

**Review by kailiangz1 — CHANGES_REQUESTED** [review]

File: drivers/common/virtio/virtio_admin.h (hunk 1 above)
Comment (GavinLi-NV) [reply drivers/common/virtio/virtio_admin.h]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 3 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio/virtio_vdpa.c]: Done

File: drivers/vdpa/virtio_mi/lm.c (hunk 5 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: Done

File: drivers/vdpa/virtio_mi/lm.c (hunk 6 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: Done

File: drivers/vdpa/virtio_mi/lm.c (hunk 7 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: Done

File: drivers/vdpa/virtio_mi/lm.c (hunk 8 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: Done

File: drivers/vdpa/virtio_mi/lm.c (hunk 9 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: What about blk device?

Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: Done

File: drivers/vdpa/virtio_mi/lm.c [hunk 10]
```diff
@@ -41,6 +41,213 @@ TAILQ_HEAD(virtio_vdpa_mi_privs, virtio_vdpa_pf_priv) virtio_mi_priv_list =
...
+
+static int
+virtio_vdpa_admin_queue_alloc(struct virtio_vdpa_pf_priv *priv)
+{
+	struct virtio_hw *hw = &priv->vpdev->hw;
+	uint16_t i, queue_idx;
+	int ret;
+
+	hw->max_queue_pairs = virtio_pci_dev_nr_vq_get(priv->vpdev) / 2;
+	priv->hw_nr_virtqs = 1;
```
Comment (kailiangz1) [inline drivers/vdpa/virtio_mi/lm.c]: use callback function to net and blk. get admin queue indx.
then init admin queue

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 2 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio/virtio_vdpa.c]: Done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 4 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio/virtio_vdpa.c]: Done

File: drivers/vdpa/virtio_mi/lm.c (hunk 10 above)
Comment (GavinLi-NV) [reply drivers/vdpa/virtio_mi/lm.c]: Done

File: drivers/common/virtio_mi/lm.c [hunk 11]
```diff
@@ -0,0 +1,1177 @@
...
+	TAILQ_ENTRY(virtio_vdpa_pf_priv) next;
+	struct rte_pci_device *pdev;
+	struct virtio_pci_dev *vpdev;
+	struct virtio_vdpa_dev_ops *dev_ops;
+	uint64_t device_features;
+	int vfio_dev_fd;
+	uint16_t hw_nr_virtqs; /* number of vq device supported*/
+};
+
+struct sge_iova {
```
Comment (kailiangz1) [inline drivers/common/virtio_mi/lm.c:47]: align to spec
diff --git a/content.tex b/content.tex
index c1cb60c..d82d399 100644
--- a/content.tex
+++ b/content.tex
@@ -2116,6 +2116,7 @@ \subsubsection{Dirty page track ctrl}\label{sec:Virtio Transport Options / Virti
struct virtio_sge {
         le64 addr;
         le32 len;
+        le32 reserved;
};

File: drivers/common/virtio_mi/lm.c [hunk 12]
```diff
@@ -0,0 +1,1178 @@
...
+
+	return 0;
+}
+
+int
+virtio_vdpa_cmd_resume(struct virtio_vdpa_pf_priv *priv, uint16_t vdev_id,
+		enum virtio_internal_status status)
+{
+	struct virtio_hw *hw;
+
```
Comment (kailiangz1) [inline drivers/common/virtio_mi/lm.c:310]: why not expose set status, suspend and resume had no different

File: drivers/common/virtio_mi/lm.c [hunk 13]
```diff
@@ -0,0 +1,1178 @@
...
+	if (!virtio_with_feature(hw, VIRTIO_F_ADMIN_VQ)) {
+		CMD_LOG(INFO, "host does not support admin queue");
+		return -ENOTSUP;
+	}
+
+	return virtio_vdpa_cmd_set_status(hw, vdev_id, status);
+}
+
+int
+virtio_vdpa_cmd_suspend(struct virtio_vdpa_pf_priv *priv, uint16_t vdev_id,
```
Comment (kailiangz1) [inline drivers/common/virtio_mi/lm.c]: we should also have get status API
VIRTIO_ADMIN_PCI_MIGRATION_GET_INTERNAL_STATUS

**Review by kailiangz1 — APPROVED** [review]

**Pull request merged**